src/main/java/org/apache/sling/junit/annotations/SlingAnnotationsTestRunner.java

```diff
@@ -54,19 +60,23 @@ protected Object createTest() throws Exception {
             return super.createTest();
         } else { 
             log.debug("Using TestObjectProcessor {}", top);
-            return top.process(super.createTest());
+            Object test = top.process(super.createTest());
+            tests.add(test);
+            return test;
         }
     }
 
     @Override
     public void run(RunNotifier notifier){
         try {
             super.run(notifier);
-        } catch (Exception e) {
-            log.error("Test 'run' method", e);
         } finally {
-            AnnotationsProcessor ap = (AnnotationsProcessor) top;
-            ap.closeAllServices();
+            AnnotationsProcessor ap = (AnnotationsProcessor) this.top;
```

Comment:
Given that there are probably no other implementations out there, and it is feasible to achieve the same with `WeakReference` + `ReferenceQueue`, I am ok with casting here. An `instanceof` check wouldn't hurt though, just in case someone dos register another `TestObjectProcessor`.